crates/oxc_transformer/src/es2022/class_static_block.rs

```diff
@@ -0,0 +1,98 @@
+use oxc_ast::{ast::*, AstBuilder};
+use oxc_span::{Atom, Span};
+
+use std::{collections::HashSet, ops::DerefMut, rc::Rc};
+
+/// ES2022: Class Static Block
+///
+/// References:
+/// * <https://babel.dev/docs/babel-plugin-transform-class-static-block>
+/// * <https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-class-static-block>
+pub struct ClassStaticBlock<'a> {
+    ast: Rc<AstBuilder<'a>>,
+}
+
+impl<'a> ClassStaticBlock<'a> {
+    pub fn new(ast: Rc<AstBuilder<'a>>) -> Self {
+        Self { ast }
+    }
+
+    pub fn transform_class_body<'b>(&mut self, class_body: &'b mut ClassBody<'a>) {
+        if !class_body.body.iter().any(|e| matches!(e, ClassElement::StaticBlock(..))) {
+            return;
+        }
+
+        let private_names: HashSet<Atom> = class_body
+            .body
+            .iter()
+            .filter_map(ClassElement::property_key)
+            .filter_map(PropertyKey::private_name)
+            .collect();
+
+        let mut i = 0;
+        for element in class_body.body.iter_mut() {
+            let ClassElement::StaticBlock(block) = element else {
+                continue;
+            };
+
+            let static_block_private_id = generate_uid(&private_names, &mut i);
+            let key = PropertyKey::PrivateIdentifier(self.ast.alloc(PrivateIdentifier {
+                span: Span::default(),
+                name: static_block_private_id.clone(),
+            }));
+
+            let value = (block.body.len() == 1
+                && matches!(block.body[0], Statement::ExpressionStatement(..)))
+            .then(|| {
+                // We special-case the single expression case to avoid the iife, since it's common.
+                let stmt = self.ast.move_statement(&mut block.body.deref_mut()[0]);
+                let Statement::ExpressionStatement(mut expr_stmt) = stmt else { unreachable!() };
+                self.ast.move_expression(&mut expr_stmt.expression)
+            })
+            .unwrap_or_else(|| {
+                let statements = self.ast.move_statement_vec(&mut block.body);
+                let callee = self.ast.parenthesized_expression(
+                    Span::default(),
+                    self.ast.arrow_expression(
+                        Span::default(),
+                        false,
+                        false,
+                        false,
+                        self.ast.formal_parameters(
```

Comment:
nit: Inlining everything makes the code harder to read - this becomes a vertical block. I usually declare a new variable for this, i.e. `let parameters = ...`, `let function_body = ...`